packages/components/src/composite/test/index.tsx

```diff
@@ -0,0 +1,536 @@
+/**
+ * External dependencies
+ */
+import { render, screen } from '@testing-library/react';
+import userEvent from '@testing-library/user-event';
+
+/**
+ * Internal dependencies
+ */
+import {
+	Composite as ReakitComposite,
+	CompositeGroup as ReakitCompositeGroup,
+	CompositeItem as ReakitCompositeItem,
+	useCompositeState as ReakitUseCompositeState,
+} from '..';
+
+const COMPOSITE_SUITES = {
+	reakit: {
+		Composite: ReakitComposite,
+		CompositeGroup: ReakitCompositeGroup,
+		CompositeItem: ReakitCompositeItem,
+		useCompositeState: ReakitUseCompositeState,
+	},
+};
+
+type InitialState = Parameters< typeof ReakitUseCompositeState >[ 0 ];
+
+// It was decided not to test the full API, instead opting
+// to cover basic usage, with a view to adding broader support
+// for the original API should the need arise. As such we are
+// only testing here for standard usage.
+// See https://github.com/WordPress/gutenberg/pull/56645
+
+describe.each( Object.entries( COMPOSITE_SUITES ) )(
+	'Validate %s implementation',
+	( _, { Composite, CompositeGroup, CompositeItem, useCompositeState } ) => {
+		function useSpreadProps( initialState?: InitialState ) {
+			return useCompositeState( initialState );
+		}
+
+		function useStateProps( initialState?: InitialState ) {
+			return {
+				state: useCompositeState( initialState ),
+			};
+		}
+
+		function useCustomProps( initialState?: InitialState ) {
+			const state = useCompositeState( initialState );
+			const { up, down, previous, next, move } = state;
+
+			return {
+				...state,
+				up: jest.fn( up ),
+				down: jest.fn( down ),
+				previous: jest.fn( previous ),
+				next: jest.fn( next ),
+				move: jest.fn( move ),
+			};
+		}
```

Comment:
I initially thought that the "custom props" scenario involved passing, as props, the same props that could be passed to the `useCompositeState` hook — if that was the case, implementing backward compatibility would be potentially be as easy as doing

```jsx
const Composite = ( {
  // Destructure all supported "custom props" 
  initialStateProp1,
  initialStateProp2,
  // state and store objects
  state,
  store,
  ...restProps, 
} ) => {
  const internalState = useCompositeState( { initialStateProp1, initialStateProp2 } );

  return (
    <Composite
      store={ store ?? state.store ?? internalState }
      { ...restProps }
    />
  );
}
```

Instead, it looks like those "custom props" are the same as the return value of the `useCompositeState` hook. If that's the case, I'm not sure how we'll be able to support backward-compat for this scenario without writing a lot of custom code, and we may need to drop this scenario in the test

Reply:
Yeah, essentially `useCompositeState` returns a collection of functions and values that components can use to read and update the shared composite state.

> These props are returned by the state hook. You can spread them into this component `({...state})` or pass them separately. You can also provide these props from your own state logic.

So there's actually nothing special about what `useCompositeState` returns, beyond it being the easiest way to make sure the state is properly shared. Any or all of the returned props could be created outside of that call.

Reply:
Got it. So, it looks like we wouldn't be able to guarantee backward compatibility on this front, unless we added explicit code supporting each of these props on `Composite` ?

Reply:
Yes. Fortunately there aren't that many, so it wouldn't be an insurmountable challenge. "Worth doing?" is of course a different question!

Reply:
I say we don't support this scenario, since we agreed that (at least in the first iteration of adding backwards-compat) we wouldn't add extra code to re-implement reakit functionality, and we would go with a minimal solution instead.

We can probably drop the `useCustomProps` scenario in the tests too, then.